hotfix/myFinalsSchedule -> dev

ecbrady opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)
<!-- Reviewable:end -->

romrell4 commented:
Reviewed 10 of 14 files at r1.
Review status: 10 of 14 files reviewed at latest revision, all discussions resolved.

---

*[a discussion](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBZc4dk1s0yZOiuirQK:-LBZc4dl3psOD_AumiLp:b-4q5ldu) (no related file):*
Does this fix the bug that happens on some iOS 11 devices? Check with Alex if you don't know what bug I'm talking about. It's probably on the VTB if you want to move it to implementing as well

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 4 of 14 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 49 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa5z9H01qnMOzwKjFa:-LBa5z9H01qnMOzwKjFb:b7nsats) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L49)):*
> ```Swift
> 
> 	func tabBarController(_ tabBarController: UITabBarController, shouldSelect viewController: UIViewController) -> Bool {
> 		if finals?.count != 0, buildings?.count != 0 {
> ```

This feels more like a guard let to me

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 84 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa6MZs48Obh3QNT0Bl:-LBa6MZs48Obh3QNT0Bm:b-fhurvt) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L84)):*
> ```Swift
> 	func createButtonFailed(button: SemesterButton2) {
> 		spinner.stopAnimating()
> 		super.displayAlert(title: "Error", message: "Unable to Load Year Terms")
> ```

I think "Error" might be the default... Double check

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 94 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa6fO018poq0F7OMuY:-LBa6fO018poq0F7OMuZ:b-dpn7jg) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L94)):*
> ```Swift
> 	
> 	private func createSemesterButton() {
> 		semButton = SemesterButton2(delegate: self)
> ```

inline this function. Unless it's used more than once (which I didn't see at first glance)

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 104 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa7EuI0HV_lWO-ez8r:-LBa7EuI0HV_lWO-ez8s:b-ptju1z) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L104)):*
> ```Swift
> 		MyFinalsScheduleClient.getFinalsExams(yearTerm: yearTerm) { (finalExams, error) in
> 			if let finalExams = finalExams {
> 				self.finals = finalExams
> ```

Just a thought, but one thing I've done in my other project is to have the didSet of `finals` and `buildings` both call dataFinishedLoading. That way you don't have to worry about calling it every time. Thoughts?

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 127 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa8UiC0y6HC77TXDRt:-LBa8UiC0y6HC77TXDRu:b-2irmih) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L127)):*
> ```Swift
> 			
> 			//Separate the finals into arrays for scheduled and unscheduled
> 			//TODO: Note to Eric: This can easily be simplified using higher order filters but would triple the runtime complexity. Your call.
> ```

Good point. How about a `removeWhere` extension that mutates the array that you're working on based on a closure, and returns the items that it removed? There's probably other ways as well if you think about it. Another one would be that you could create an extension `splitOn` and pass in a variable number of closures, and it will return you a list of that size plus one (defaulting to the "else"). Let me know what you think would be best

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 19 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaVTUJ2a0Te124SCyM:-LBaVTUJ2a0Te124SCyN:blyiyy6) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L19)):*
> ```Swift
> 	//MARK: Private Properties
> 	
> 	private var locationManager = CLLocationManager()
> ```

I don't think we need a location manager here. If they have given authorization, their dot will show up on the map. If they haven't, it's not worth asking them for it.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 34 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaVox4CfOJGWz0fGdv:-LBaVox4CfOJGWz0fGdw:b-b6uw7c) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L34)):*
> ```Swift
> 		
> 		//If the user is coming to the mapView by selecting a final, we want to select the annotation associated with that final
> 		if let final = selectedFinal, let selectedBuilding = mapView.annotations.first(where: { (annotation) -> Bool in
> ```

Let's save the `selectedBuilding` on the previous line so that the if statement is easier to read

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 35 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaW-4X3pR4qW-twuku:-LBaW-4X3pR4qW-twukv:b-n6u24v) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L35)):*
> ```Swift
> 		//If the user is coming to the mapView by selecting a final, we want to select the annotation associated with that final
> 		if let final = selectedFinal, let selectedBuilding = mapView.annotations.first(where: { (annotation) -> Bool in
> 			if let building = annotation as? CampusBuilding2 {
> ```

If do we need the if/let? Will optional chaining not work? If it will, let's use it and shorthand notation (I think, unless you think it looks convoluted with it)

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 62 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaWFdfCEOZFZYkVB32:-LBaWFdfCEOZFZYkVB33:bkldqrx) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L62)):*
> ```Swift
> 		mapView.addAnnotations(finalsBuildings)
> 		
> 		zoomToFitMapAnnotations(includeUserLocation: false, keepByuInFrame: true)
> ```

Yeah, if we don't even zoom to their dot, no use in asking for their location to be turned on

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 73 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaWQ4ACicJE12RVt3Z:-LBaWQ4ACicJE12RVt3_:b-usiaa5) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L73)):*
> ```Swift
> 			} else {
> 				//No annotation view exists in the reuse queue; create one
> 				return CampusLocationsHelper2.pinAnnotationView(hideDetailDisclosureButton: false, forBuildingAnnotation: building, animatingDrop: false)
> ```

Instead of doing this in a helper, could we make this an extension on the MapView? I want to get away from having "helpers" now that we have extensions

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 75 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaWeENEmNAtFT9Nhdy:-LBaWeENEmNAtFT9Nhdz:b-oyhbfg) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L75)):*
> ```Swift
> 				return CampusLocationsHelper2.pinAnnotationView(hideDetailDisclosureButton: false, forBuildingAnnotation: building, animatingDrop: false)
> 			}
> 		} else {
> ```

In fact... We do this logic so often, maybe this entire function should be an extension on MapView... Tell me what you think

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 114 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaX30y2vA1xjBrvVSQ:-LBaX30y2vA1xjBrvVSR:bnogvb5) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L114)):*
> ```Swift
> 		var firstFinal = true
> 		
> 		for final in finals {
> ```

This could (and should) be done with a reduce. It will handle the first one not having "\n", etc. Let me know if you hit snags trying to implement it. You might have to do a filter first to get only the ones in that building, but that shouldn't be a problem

---

*[byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift, line 9 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaXKYb6Gc5WNa4eOMT:-LBaXKYb6Gc5WNa4eOMU:btkeqzv) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift#L9)):*
> ```Swift
> //
> 
> class FinalExam: NSObject {
> ```

Do we need to be an NSObject? If not, remove it

---

*[byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift, line 60 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaXZowD_m_YZqsC6Tm:-LBaXZowD_m_YZqsC6Tn:b270a1l) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c251ab1c9df6b34d90e2e31a3606b0e38ea9566b/byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift#L60)):*
> ```Swift
> 	
> 	func testLocationDetailsText() -> String {
> 		var detaiText = ""
> ```

I believe this could be done with a reduce. Let me know if you have troubles

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

ecbrady commented:
Review status: 11 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBZc4dk1s0yZOiuirQK:-LBa_Ho_AbnWTxff56lH:b8m2niu) (no related file):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Does this fix the bug that happens on some iOS 11 devices? Check with Alex if you don't know what bug I'm talking about. It's probably on the VTB if you want to move it to implementing as well
</blockquote></details>

Yes it does. It's funny because I found that bug the same day you did but then we both totally forgot until I was having Geoff test it and he found it so I fixed it. It was just a small issue with the storyboard.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 49 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa5z9H01qnMOzwKjFa:-LBvIa1TE845IWbFr-Km:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L49)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

This feels more like a guard let to me
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 84 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa6MZs48Obh3QNT0Bl:-LBvIfSV8ddVitihmxaP:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L84)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I think "Error" might be the default... Double check
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 94 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa6fO018poq0F7OMuY:-LBvInqX7DimgHSLvoud:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L94)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

inline this function. Unless it's used more than once (which I didn't see at first glance)
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 104 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa7EuI0HV_lWO-ez8r:-LBvJEIv1O_khNDJHY9z:b-utdjnf) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L104)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Just a thought, but one thing I've done in my other project is to have the didSet of `finals` and `buildings` both call dataFinishedLoading. That way you don't have to worry about calling it every time. Thoughts?
</blockquote></details>

I like that but it adds lines of code in this case. I think it would make sense if we were loading each individually in multiple places but since each of these is only loaded in one place then either we call it only once per variable.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 127 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa8UiC0y6HC77TXDRt:-LBwC9ch37eMDCOVmO-2:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L127)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Good point. How about a `removeWhere` extension that mutates the array that you're working on based on a closure, and returns the items that it removed? There's probably other ways as well if you think about it. Another one would be that you could create an extension `splitOn` and pass in a variable number of closures, and it will return you a list of that size plus one (defaulting to the "else"). Let me know what you think would be best
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 19 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaVTUJ2a0Te124SCyM:-LBvTYfT4BgcGOkgBFLl:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L19)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I don't think we need a location manager here. If they have given authorization, their dot will show up on the map. If they haven't, it's not worth asking them for it.
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 34 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaVox4CfOJGWz0fGdv:-LBvTtuH3B6F4tRf3hsG:bhhumwk) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L34)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Let's save the `selectedBuilding` on the previous line so that the if statement is easier to read
</blockquote></details>

We can't because the closure depends on the final which we if let before setting selectedBuilding. The alternative is to nest if lets.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 35 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaW-4X3pR4qW-twuku:-LBvU4875FTEe6-F3Ulb:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L35)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

If do we need the if/let? Will optional chaining not work? If it will, let's use it and shorthand notation (I think, unless you think it looks convoluted with it)
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 62 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaWFdfCEOZFZYkVB32:-LBvUXab5_tW0lWozp_n:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L62)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Yeah, if we don't even zoom to their dot, no use in asking for their location to be turned on
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 114 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaX30y2vA1xjBrvVSQ:-LBwBkYTBhFI7vlDHa84:bln777q) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L114)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

This could (and should) be done with a reduce. It will handle the first one not having "\n", etc. Let me know if you hit snags trying to implement it. You might have to do a filter first to get only the ones in that building, but that shouldn't be a problem
</blockquote></details>

Done. It became so simple I just got rid of the method entirely. 

---

*[byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift, line 9 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaXKYb6Gc5WNa4eOMT:-LBwBx3B-8LmTrzirsq_:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift#L9)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Do we need to be an NSObject? If not, remove it
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift, line 60 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaXZowD_m_YZqsC6Tm:-LBwC4zrDsiiLsjdI1Uj:bm76s3e) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift#L60)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I believe this could be done with a reduce. Let me know if you have troubles
</blockquote></details>

I don't think so since these are coming from different variables and not one array of values.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

ecbrady commented:
Review status: 11 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 73 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaWQ4ACicJE12RVt3Z:-LBwDswq4qS-q8oprArq:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L73)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Instead of doing this in a helper, could we make this an extension on the MapView? I want to get away from having "helpers" now that we have extensions
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 75 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaWeENEmNAtFT9Nhdy:-LBwHnTu3oO6xRAyGunD:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L75)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

In fact... We do this logic so often, maybe this entire function should be an extension on MapView... Tell me what you think
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 4 of 4 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[byuSuite/Apps/MapMyClasses/controller/MapMyClassesMapViewController.swift, line 64 at r3](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwJlBU2u3bNzT9m97I:-LBwJlBU2u3bNzT9m97J:bjne2p2) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/aeaff55e8f7dd7a7753a90a3f45014ee9f99d1a1/byuSuite/Apps/MapMyClasses/controller/MapMyClassesMapViewController.swift#L64)):*
> ```Swift
> 	
> 	func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
> 		return mapView.defaultBuildingAnnotationView(annotation: annotation, hidingDetailDisclousreButton: false)
> ```

Should we default hidingDetailDisclosure to something? What is the most common case? Also, why "hiding" instead of "displayDetailDisclosureButton"? If no reason, change to the other

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 49 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa5z9H01qnMOzwKjFa:-LBwGPkr4nBMqVugSak9:b716kgu) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L49)):*
<details><summary><i>Previously, ecbrady (Erik Brady) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Oh you can still use the optional chaining if you want. You didn't have to let the variables first in the guard. You can just use
```swift
guard finals?.count != 0, buildings?.count != 0 else { return false }
```

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 34 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaVox4CfOJGWz0fGdv:-LBwHcLK2-az90RBFRhi:b-qwzxoj) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L34)):*
<details><summary><i>Previously, ecbrady (Erik Brady) wrote…</i></summary><blockquote>

We can't because the closure depends on the final which we if let before setting selectedBuilding. The alternative is to nest if lets.
</blockquote></details>

Do we really need `final`? Can we not just use optional chaining in the closure?

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 114 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaX30y2vA1xjBrvVSQ:-LBwHvb--IovywRy7_oO:b1mhcef) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L114)):*
<details><summary><i>Previously, ecbrady (Erik Brady) wrote…</i></summary><blockquote>

Done. It became so simple I just got rid of the method entirely. 
</blockquote></details>

Hey hey! That looks WAY nice! Good job!

---

*[byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift, line 60 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaXZowD_m_YZqsC6Tm:-LBwI2u49oPuVHOtDZ_y:b-eai2gd) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/model/FinalExam.swift#L60)):*
<details><summary><i>Previously, ecbrady (Erik Brady) wrote…</i></summary><blockquote>

I don't think so since these are coming from different variables and not one array of values.
</blockquote></details>

Yeah... it could still be done if you did this:
```swift
[building, room, days, classPeriod].reduce(....)
```
But now that I look at it again, this looks like the best way of doing it

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 669 at r2](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwIh203nZMq7ySmPnb:-LBwIh203nZMq7ySmPnc:b-w2t5e3) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/d280cb7f8268a294ffdb0b43ef90bcdb6c3902ef/byuSuite/Classes/Reusable/Extensions.swift#L669)):*
> ```Swift
> 
> extension Array {
> 	mutating func filterAndRemove(byClosure: (Element) -> Bool) -> [Element] {
> ```

rename to `filterAndRemove(where: (Element) -> Bool)`

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 670 at r2](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwIW7nE_kF9Y-cqDAN:-LBwIW7nE_kF9Y-cqDAO:b-pbyq75) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/d280cb7f8268a294ffdb0b43ef90bcdb6c3902ef/byuSuite/Classes/Reusable/Extensions.swift#L670)):*
> ```Swift
> extension Array {
> 	mutating func filterAndRemove(byClosure: (Element) -> Bool) -> [Element] {
> 		var filteredElements = [Element]()
> ```

This looks good.

However, it would be nice if we didn't have to reset self at the end... What if we just removed from self and added to a `result`. That way we don't have to store multiple copies of the same data. Look into it. But if it gets too complicated, this solution is totally fine for our use case

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 715 at r3](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwJ9L0EkkeP_w4pePw:-LBwJ9L0EkkeP_w4pePx:bs6irh) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/aeaff55e8f7dd7a7753a90a3f45014ee9f99d1a1/byuSuite/Classes/Reusable/Extensions.swift#L715)):*
> ```Swift
> 	
> 	func defaultBuildingAnnotationView(annotation: MKAnnotation, hidingDetailDisclousreButton: Bool) -> MKAnnotationView? {
> 		guard let building = annotation as? CampusBuilding2 else {
> ```

If the guard statement is short, I like seeing it as a one liner `guard ... else { return nil }`. Then the comment can go above the guard. I think it makes the guard easier to differentiate from the regular logic (also, always put a newline between the guard and the subsequent logic)

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 736 at r3](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwJY_z2UkcVYnaRYk3:-LBwJY_z2UkcVYnaRYk4:b-at4vz6) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/aeaff55e8f7dd7a7753a90a3f45014ee9f99d1a1/byuSuite/Classes/Reusable/Extensions.swift#L736)):*
> ```Swift
> 			//Add detail button to callout
> 			pinView.rightCalloutAccessoryView = UIButton(type: .detailDisclosure)
> 		}
> ```

Do we need an else here? It seems that if it recycles one, the rightCalloutAccessoryView might still exist from the last one if we don't go into the if

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

ecbrady commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 669 at r2](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwIh203nZMq7ySmPnb:-LBwWYigCTmHY2WJH9Xk:bgo3w36) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/d280cb7f8268a294ffdb0b43ef90bcdb6c3902ef/byuSuite/Classes/Reusable/Extensions.swift#L669)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

rename to `filterAndRemove(where: (Element) -> Bool)`
</blockquote></details>

I would need a variable name since where is a keyword. So like (where variable: (Ele...

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 670 at r2](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwIW7nE_kF9Y-cqDAN:-LBwVyjZDJcRxsKU2rQ_:b-qwfhdv) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/d280cb7f8268a294ffdb0b43ef90bcdb6c3902ef/byuSuite/Classes/Reusable/Extensions.swift#L670)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

This looks good.

However, it would be nice if we didn't have to reset self at the end... What if we just removed from self and added to a `result`. That way we don't have to store multiple copies of the same data. Look into it. But if it gets too complicated, this solution is totally fine for our use case
</blockquote></details>

The only reason I did it this way was to avoid crashes. Correct me if I'm wrong, but to remove it we would have to remove at index and by removing it it would throw off the loop.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

ecbrady commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[byuSuite/Apps/MapMyClasses/controller/MapMyClassesMapViewController.swift, line 64 at r3](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwJlBU2u3bNzT9m97I:-LBwX2g1AlXHwSac4DgR:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/aeaff55e8f7dd7a7753a90a3f45014ee9f99d1a1/byuSuite/Apps/MapMyClasses/controller/MapMyClassesMapViewController.swift#L64)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Should we default hidingDetailDisclosure to something? What is the most common case? Also, why "hiding" instead of "displayDetailDisclosureButton"? If no reason, change to the other
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift, line 49 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBa5z9H01qnMOzwKjFa:-LBwXGkH--czg1cPMTWq:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleListViewController.swift#L49)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Oh you can still use the optional chaining if you want. You didn't have to let the variables first in the guard. You can just use
```swift
guard finals?.count != 0, buildings?.count != 0 else { return false }
```
</blockquote></details>

Done.

---

*[byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift, line 34 at r1](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBaVox4CfOJGWz0fGdv:-LBwXa2q-0Lpynv-Fxtr:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fb02d72e4c23adb162c617c85a7668b1c53aeecb/byuSuite/Apps/MyFinalsSchedule/controller/MyFinalsScheduleMapViewController.swift#L34)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Do we really need `final`? Can we not just use optional chaining in the closure?
</blockquote></details>

Done.

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 715 at r3](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwJ9L0EkkeP_w4pePw:-LBwXkdJ5ecHeT4U519Q:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/aeaff55e8f7dd7a7753a90a3f45014ee9f99d1a1/byuSuite/Classes/Reusable/Extensions.swift#L715)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

If the guard statement is short, I like seeing it as a one liner `guard ... else { return nil }`. Then the comment can go above the guard. I think it makes the guard easier to differentiate from the regular logic (also, always put a newline between the guard and the subsequent logic)
</blockquote></details>

Done.

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 736 at r3](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwJY_z2UkcVYnaRYk3:-LBwY3G852InsNA6vHN2:b-oz8tym) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/aeaff55e8f7dd7a7753a90a3f45014ee9f99d1a1/byuSuite/Classes/Reusable/Extensions.swift#L736)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Do we need an else here? It seems that if it recycles one, the rightCalloutAccessoryView might still exist from the last one if we don't go into the if
</blockquote></details>

Good call.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[byuSuite/Apps/MyClasses/controller/MyClassesDetailViewController.swift, line 51 at r4](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBzIhKp-IzWOQuqyEQd:-LBzIhKp-IzWOQuqyEQe:bqe02s) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fe3487f2738502a8eda7306a0a805ddaf38b1f87/byuSuite/Apps/MyClasses/controller/MyClassesDetailViewController.swift#L51)):*
> ```Swift
> 	
> 	func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
> 		return mapView.defaultBuildingAnnotationView(annotation: annotation, displayDetailDisclousreButton: true)
> ```

So before we were hiding the button... are we now showing it? (We changed the name of the parameter to the opposite `hide`->`display`, but didn't change the passed in value from `true`->`false`)

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 669 at r2](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwIh203nZMq7ySmPnb:-LBzJ4LF14L6sTTTcCyG:bjvsxpg) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/d280cb7f8268a294ffdb0b43ef90bcdb6c3902ef/byuSuite/Classes/Reusable/Extensions.swift#L669)):*
<details><summary><i>Previously, ecbrady (Erik Brady) wrote…</i></summary><blockquote>

I would need a variable name since where is a keyword. So like (where variable: (Ele...
</blockquote></details>

Oh yeah. Dang. You pick the variable name. Maybe something like: `filterAndRemove(where closure: (Element) -> Bool)`

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 714 at r4](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBzJVofCJrHQxJHMdsT:-LBzJVofCJrHQxJHMdsU:b-5neki7) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fe3487f2738502a8eda7306a0a805ddaf38b1f87/byuSuite/Classes/Reusable/Extensions.swift#L714)):*
> ```Swift
> 	}
> 	
> 	func defaultBuildingAnnotationView(annotation: MKAnnotation, displayDetailDisclousreButton: Bool = false) -> MKAnnotationView? {
> ```

typo in disclosure

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 725 at r4](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBzJc1A0O6gq_by85Z6:-LBzJc1A0O6gq_by85Z7:b-jkls3z) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/fe3487f2738502a8eda7306a0a805ddaf38b1f87/byuSuite/Classes/Reusable/Extensions.swift#L725)):*
> ```Swift
> 	}
> 	
> 	func pinAnnotationView(hideDetailDisclosureButton: Bool, forBuildingAnnotation annotation: CampusBuilding2, animatingDrop: Bool) -> MKPinAnnotationView {
> ```

change hide to display and make sure the if statement below is changed as well (remove the `!`)

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

ecbrady commented:
Review status: 15 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[byuSuite/Classes/Reusable/Extensions.swift, line 669 at r2](https://reviewable.io/reviews/byu-oit/ios-byusuite/334#-LBwIh203nZMq7ySmPnb:-LC-aYXl0vbUKUh-8EHF:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/d280cb7f8268a294ffdb0b43ef90bcdb6c3902ef/byuSuite/Classes/Reusable/Extensions.swift#L669)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Oh yeah. Dang. You pick the variable name. Maybe something like: `filterAndRemove(where closure: (Element) -> Bool)`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/ios-byusuite/334)*
<!-- Sent from Reviewable.io -->
